src/server.ts

```diff
@@ -749,3 +752,68 @@ function findCreatedAccountSequenceInTransactionMeta(
 
   throw new Error("No account created in transaction");
 }
+
+
+// TODO - remove once rpc updated to
+// append expiration entry per data LK requested onto server-side response 
+// https://github.com/stellar/soroban-tools/issues/1010
+function mergeResponseExpirationLedgers(ledgerEntriesResponse: SorobanRpc.RawGetLedgerEntriesResponse,
+  requestedKeys: xdr.LedgerKey[]
+): SorobanRpc.RawGetLedgerEntriesResponse {
+  const expirationKeyToRawEntryResult = new Map<String, SorobanRpc.RawLedgerEntryResult>();
+  const requestedKeyXdrs = new Set<String>(requestedKeys.map(requestedKey =>
+    requestedKey.toXDR('base64')));
+
+  (ledgerEntriesResponse.entries ?? []).forEach(rawEntryResult => {
+    if (!rawEntryResult.key || !rawEntryResult.xdr) {
+      throw new TypeError(`invalid ledger entry: ${JSON.stringify(rawEntryResult)}`);
+    }
+    
+    const parsedKey = xdr.LedgerKey.fromXDR(rawEntryResult.key, 'base64');
+    
+    if (parsedKey.switch().value !== xdr.LedgerEntryType.expiration().value ||
```

Comment:
right okay so I looked at this carefully again and what you want to do is:

- if it's a real ledger entry (not expiration) and key exists, merge with existing entry
- if key not exists, add new entry (w/ key, xdr)
- if an expiration entry, merge or add also

But all of this can be ultra-simplified into just "merge existing (if any) with key hash no matter what":

```typescript
const isExpiration = parsedKey.switch().value === xdr.LedgerEntryType.expiration().value;
const keyHash = (isExpiration) 
  ? parsedKey.expiration().keyHash().toString()
  : hash(parsedKey.toXDR()).toString();

const existing = expirationKeyToRawEntryResult.get(keyHash) ?? {};

if (isExpiration) {
  const expiration = xdr.LedgerEntryData
      .fromXDR(rawEntryResult.xdr, 'base64')
      .expiration()
      .toXDR('base64');
  expirationKeyToRawEntryResult.set(keyHash, { ...existing, expiration });
} else {
  expirationKeyToRawEntryResult.set(keyHash, { ...existing, ...rawEntryResult });
}
```

Reply:
yes, great optimization, squeaked it in here https://github.com/stellar/js-soroban-client/pull/161/commits/3b66c37bf87a02c065b4e0aa6e909feed45b46fd